### ISSUE TITLE
Lint for cast to inline class

### ISSUE THREAD

**Issue by Cat-sushi** [dart-lang/language#3180]
I understand that `as` is a responsibility delegation from the compiler to users.
And, the compiler doesn't check down casts with `as`, but the runtime does.

On the other hand, the runtime skip user-defined check logic on constructor of `inline class`.

So, I propose a new lint to warn a cast from the representation type to the `inline class` with check logic or body on its constructor, without implicit constructor.

**Comment by eernstg** [comment]
Right, we've discussed having this kind of lint as well. It should apply in the case where any expression is subject to a cast to an inline type, `e is InlineType`, but probably also for ary cast where this kind of change is a consequence of the structure: `myListOfRepresentationType as List<InlineType>`, or casting a `void Function(InlineType)` to `void Function(RepresentationType)` (because this cast will allow us to silently turn an expression of `RepresentationType` into an `InlineType` every time we call that function).

On the other hand, if the inline class can be declared to be a subtype of a non-inline class then the authors/maintainers of the `InlineClass` have already decided that it should be very easy to go from one to the other and back. In that case we would presumably not emit any lints.
